File: typescript-examples/network-interception/api/network-interceptor.ts
import { Page, BrowserContext, Request } from "playwright";
import { goToUrl, withNetworkSettledWait } from "@intuned/browser";
import { attemptStore } from "@intuned/runtime";
import { z } from "zod";
import {
  paramsSchema,
  consultationSchema,
} from "../utils/typesAndSchemas";

type Params = z.infer<typeof paramsSchema>;
type Consultation = z.infer<typeof consultationSchema>;

const SUPABASE_URL = "https://pfaqkmnqacbiimndjwyp.supabase.co";

async function login(
  page: Page,
  email: string,
  password: string
): Promise<void> {
  const loginUrl = "https://sandbox.intuned.dev/login";
  console.log(`Navigating to login page: ${loginUrl}`);
  await goToUrl({ page, url: loginUrl });
  await page.locator("input[type='email']").fill(email);
  await page.locator("input[type='password']").fill(password);
  await page.locator("button[type='submit']").click();
  console.log("Login form submitted");
  await page.waitForTimeout(3000);
}

async function interceptRequest(request: Request): Promise<void> {
  if (!request.url().includes(SUPABASE_URL)) return;

  if (attemptStore.get("bearer_token") === null) {
    const authHeader = request.headers()["authorization"] ?? "";
    const userIdHeader = request.headers()["x-user-id"] ?? "";

    if (authHeader.startsWith("Bearer ")) {
      const token = authHeader.slice("Bearer ".length);
      attemptStore.set("bearer_token", token);
      console.log(`Bearer token captured: ${token.slice(0, 20)}...`);
    }

    // Extract user_id from the consultations URL pattern if present
    if (
      request.url().includes("/consultations") &&
      request.url().includes("user_id=eq.")
    ) {
      const uid = request.url().split("user_id=eq.")[1].split("&")[0];
      attemptStore.set("user_id", uid);
      console.log(`User ID captured: ${uid}`);
    } else if (userIdHeader) {
      attemptStore.set("user_id", userIdHeader);
    }
  }

  if (attemptStore.get("anon_key") === null) {
    const anonKey = request.headers()["apikey"];
    if (anonKey) {
      attemptStore.set("anon_key", anonKey);
      console.log(`Anon key captured: ${anonKey.slice(0, 20)}...`);
    }
  }
}

/**
 * Logs in to sandbox.intuned.dev, intercepts Supabase requests to capture
 * the Bearer token and anon key, then fetches consultations via the REST API.
 *
 * Example params:
 * {
 *   "username": "demo@email.com",
 *   "password": "DemoUser2024!",
 *   "limit": 5
 * }
 */
async function handler(
  params: Params,
  page: Page,
  _context: BrowserContext
): Promise<Consultation[]> {
  attemptStore.set("bearer_token", null);
  attemptStore.set("anon_key", null);
  attemptStore.set("user_id", null);

  await login(page, params.username, params.password);

  page.on("request", interceptRequest);

  try {
    console.log("Navigating to consultations page to capture API credentials...");
    await withNetworkSettledWait(
      async (page) => {
        await goToUrl({ page, url: "https://sandbox.intuned.dev/consultations-auth/list" });
      },
      { page, timeoutInMs: 20000 }
    );

    // Poll until both token and anon key are captured (up to 10 seconds)
    for (let i = 0; i < 20; i++) {
      if (attemptStore.get("bearer_token") && attemptStore.get("anon_key")) break;
      await page.waitForTimeout(500);
    }

    const bearerToken = attemptStore.get("bearer_token") as string | null;
    const anonKey = attemptStore.get("anon_key") as string | null;
    const userId = attemptStore.get("user_id") as string | null;

    if (!bearerToken) {
      throw new Error("Failed to capture Bearer token — check credentials");
    }
    if (!anonKey) {
      throw new Error("Failed to capture Supabase anon key from network");
    }

    const apiUrl =
      `${SUPABASE_URL}/rest/v1/consultations` +
      `?select=*` +
      `&user_id=eq.${userId}` +
      `&order=created_at.desc` +
      `&offset=0` +
      `&limit=${params.limit}`;

    console.log(`Fetching consultations from Supabase: ${apiUrl}`);

    const result = await page.evaluate(
      async (options: { url: string; bearerToken: string; anonKey: string }) => {
        const response = await fetch(options.url, {
          headers: {
            accept: "*/*",
            "accept-profile": "public",
            apikey: options.anonKey,
            authorization: `Bearer ${options.bearerToken}`,
            prefer: "count=exact",
          },
        });
        if (!response.ok) {
          throw new Error(`Request failed: ${response.status} ${response.statusText}`);
        }
        return response.json();
      },
      { url: apiUrl, bearerToken, anonKey }
    );

    const consultations = (result as unknown[]).map((item) =>
      consultationSchema.parse(item)
    );
    console.log(`Fetched ${consultations.length} consultations`);
    return consultations;
  } finally {
    page.removeListener("request", interceptRequest);
    console.log("Request interceptor removed");
  }
}

export default handler;

File: python-examples/network-interception/api/network-interceptor.py
from intuned_browser import go_to_url, wait_for_network_settled
from intuned_runtime import attempt_store
from playwright.async_api import BrowserContext, Page, Request
from utils.types_and_schemas import Consultation, NetworkInterceptorParams

# Supabase project base URL for sandbox.intuned.dev
SUPABASE_URL = "https://pfaqkmnqacbiimndjwyp.supabase.co"


async def login(page: Page, email: str, password: str) -> None:
    """Log in to sandbox.intuned.dev with the given credentials."""
    login_url = "https://sandbox.intuned.dev/login"
    print(f"Navigating to login page: {login_url}")
    await go_to_url(page, login_url)
    await page.locator("input[type='email']").fill(email)
    await page.locator("input[type='password']").fill(password)
    await page.locator("button[type='submit']").click()
    print("Login form submitted")
    await page.wait_for_timeout(3000)


async def intercept_request(request: Request) -> None:
    """Capture the Bearer token, anon key, and user ID from Supabase API requests."""
    if SUPABASE_URL not in request.url:
        return

    if attempt_store.get("bearer_token") is None:
        auth_header = request.headers.get("authorization", "")
        user_id_header = request.headers.get("x-user-id", "")

        if auth_header.startswith("Bearer "):
            token = auth_header[len("Bearer "):]
            attempt_store.set("bearer_token", token)
            print(f"Bearer token captured: {token[:20]}...")

        # Extract user_id from the consultations URL pattern if present
        if "/consultations" in request.url and "user_id=eq." in request.url:
            uid = request.url.split("user_id=eq.")[1].split("&")[0]
            attempt_store.set("user_id", uid)
            print(f"User ID captured: {uid}")
        elif user_id_header:
            attempt_store.set("user_id", user_id_header)

    if attempt_store.get("anon_key") is None:
        anon_key = request.headers.get("apikey")
        if anon_key:
            attempt_store.set("anon_key", anon_key)
            print(f"Anon key captured: {anon_key[:20]}...")


async def automation(
    page: Page,
    params: NetworkInterceptorParams,
    context: BrowserContext | None = None,
    **_kwargs,
) -> list[Consultation]:
    """
    Logs in to sandbox.intuned.dev, intercepts a Supabase API request to capture
    the Bearer token, then fetches consultations from the Supabase REST API.

    Example parameters:
    {
        "username": "demo@email.com",
        "password": "demo123",
        "limit": 5
    }
    """
    params = NetworkInterceptorParams(**params)
    attempt_store.set("bearer_token", None)
    attempt_store.set("anon_key", None)
    attempt_store.set("user_id", None)

    await login(page, params.username, params.password)

    # Attach interceptor and navigate to a page that triggers authenticated Supabase calls
    page.on("request", intercept_request)

    try:
        print("Navigating to consultations page to capture API token...")
        await wait_for_network_settled(
            page=page,
            func=lambda: go_to_url(page, "https://sandbox.intuned.dev/consultations-auth/list"),
            timeout_s=20,
        )

        # Poll until the token is captured (up to 10 seconds)
        for _ in range(20):
            if attempt_store.get("bearer_token"):
                break
            await page.wait_for_timeout(500)

        bearer_token = attempt_store.get("bearer_token")
        anon_key = attempt_store.get("anon_key")
        user_id = attempt_store.get("user_id")

        if not bearer_token:
            raise ValueError("Failed to capture Bearer token — check credentials or page selectors")
        if not anon_key:
            raise ValueError("Failed to capture Supabase anon key from network")

        api_url = (
            f"{SUPABASE_URL}/rest/v1/consultations"
            f"?select=*"
            f"&user_id=eq.{user_id}"
            f"&order=created_at.desc"
            f"&offset=0"
            f"&limit={params.limit}"
        )

        print(f"Fetching consultations from Supabase: {api_url}")

        result = await page.evaluate(
            """
            async (options) => {
                const { url, bearerToken, anonKey } = options;
                const response = await fetch(url, {
                    headers: {
                        "accept": "*/*",
                        "accept-profile": "public",
                        "apikey": anonKey,
                        "authorization": "Bearer " + bearerToken,
                        "prefer": "count=exact"
                    }
                });
                if (!response.ok) {
                    throw new Error("Request failed: " + response.status + " " + response.statusText);
                }
                return await response.json();
            }
            """,
            {"url": api_url, "bearerToken": bearer_token, "anonKey": anon_key},
        )

        print(f"Fetched {len(result)} consultations")
        return [Consultation(**item) for item in result]

    finally:
        page.remove_listener("request", intercept_request)
        print("Request interceptor removed")
